imap/imapd.c

```diff
@@ -873,11 +873,27 @@ static void imapd_reset(void)
     /* log the client behaviors */
     xsyslog(LOG_NOTICE, "session ended",
                         "sessionid=<%s> userid=<%s>"
-                        " compress=<%u> notify=<%u>",
+                        "%s%s%s%s"
+                        "%s%s%s%s"
+                        "%s%s%s%s",
+
                         session_id(),
                         imapd_userid ? imapd_userid : "",
-                        client_behavior.did_compress,
-                        client_behavior.did_notify);
+
+                        client_behavior.did_annotate  ? " annotate=<1>"   : "",
+                        client_behavior.did_compress  ? " compress=<1>"   : "",
+                        client_behavior.did_condstore ? " condstore=<1>"  : "",
+                        client_behavior.did_idle      ? " idle=<1>"       : "",
+
+                        client_behavior.did_metadata  ? " metadata=<1>"   : "",
+                        client_behavior.did_move      ? " move=<1>"       : "",
+                        client_behavior.did_notify    ? " notify=<1>"     : "",
+                        client_behavior.did_preview   ? " preview=<1>"    : "",
+
+                        client_behavior.did_qresync   ? " qresync=<1>"    : "",
+                        client_behavior.did_replace   ? " replace=<1>"    : "",
+                        client_behavior.did_savedate  ? " savedate=<1>"   : "",
+                        client_behavior.did_searchres ? " searchres=<1>"  : "");
```

Comment:
So, I assume there's a reason you're omitting fields that are zero, rather than just logging a lot of `"field=<%u>"` which will get zeroes or ones.  I might like to see a comment explaining the complexity.

If you don't mind rearchitecting the whole thing, you could have the computer compute the log string by hardcoding more metadata.  Something like (untested, written directly in the github web ui, probably full of typos, so think of this as pseudocode...):

```
enum client_behaviour {
    CLIENT_DID_ANNOTATE,
    // [...]
    CLIENT_DID_SEARCHRES,

    NUM_CLIENT_BEHAVIOURS /* leave last! */
};

static uint8_t client_behavior_registry[NUM_CLIENT_BEHAVIOURS];

static const char const *client_behavior_name[NUM_CLIENT_BEHAVIOURS] = {
    "annotate",
    // [...]
    "searchres",
};

static void get_client_behaviors_string(struct buf *buf)
{
    const char *sep = "";
    unsigned i;

    for (i = 0; i < NUM_CLIENT_BEHAVIOURS; i++) {
        // XXX decide whether to always log foo=<%u>, or only log when foo=<1>
        buf_printf(buf, "%s%s=<%u>",
            sep,
            client_behaviour_name[i],
            client_behaviour_registry[i]);
        sep = " ";
    }
}

```

And then each time a client does a thing, you'd do something like:

```
   client_behaviour_registry[CLIENT_DID_ANNOTATE] = 1;
```

And in `imapd_reset()`, you'd do something like:

```
    struct buf behaviors = BUF_INITIALIZER;
    // [...]

    get_client_behaviors_string(&behaviors);

    xsyslog(LOG_NOTICE, "session ended",
                        "sessionid=<%s> userid=<%s> %s"
                        session_id(),
                        imapd_userid ? imapd_userid : "",
                        buf_cstring(&behaviors));

    // [...]
    memset(client_behavior_registry, 0, sizeof client_behavior_registry);
    buf_free(&behaviours);
```

This would come at the cost of each behaviour flag taking a byte rather than a bit to store, but I still don't think that's an optimisation worth chasing.  If you wanted to chase that anyway, you could make client_behaviour_registry a large unsigned integer, and if the client performs behaviour n (according to the enum), set the nth bit.  You could also do the same with a `struct { foo :1; }`, but that would be brittle, because then you'd have three lists to have to keep coordinated (behaviour ids, behaviour names, struct field names).

Reply:
Thanks, I have added a comment.  Your provided code was interesting and provided some implicit pointers (ha) on C idiom for me.  I think for now, I'll leave things as they are, but may come back to this in the future if it seems like the trade-offs changes…